### ISSUE TITLE
With syslog, an error occured.

### ISSUE THREAD

**Issue by odg0318** [kassisol/hbm#21]
Hello.

### Summary
I have a troublesome integrating with syslog on a specific machine.

### Version
```bash
$ hbm version
Version:     0.9.2
Git commit:  878ff89
Built:       2018-03-13 05:30:07 +0900 KST
Go version:  go1.8.3
OS/Arch:     linux/amd64
```
```bash
$ docker version
Client:
 Version:	17.12.0-ce
 API version:	1.35
 Go version:	go1.9.2
 Git commit:	c97c6d6
 Built:	Wed Dec 27 20:10:14 2017
 OS/Arch:	linux/amd64

Server:
 Engine:
  Version:	17.12.0-ce
  API version:	1.35 (minimum version 1.12)
  Go version:	go1.9.2
  Git commit:	c97c6d6
  Built:	Wed Dec 27 20:12:46 2017
  OS/Arch:	linux/amd64
  Experimental:	false
```
```bash
$ docker-compose version
docker-compose version 1.17.0, build ac53b73
docker-py version: 2.5.1
CPython version: 2.7.13
OpenSSL version: OpenSSL 1.0.1t  3 May 2016
```

I used `docker-compose up` to run containers with below docker-compose file.
```yaml
version: '3'

services:
  syslog:
    image: private.registry.url/library/syslog:alpine
    volumes:
     - ./log:/var/log/docker
    ports:
     - 127.0.0.1:10514:10514

  nginx:
    image: private.registry.url/library/nginx:1.13.9-alpine
    restart: always
    ports:
     - 80:80
     - 443:443
    volumes:
     - ./nginx.conf:/etc/nginx/nginx.conf
     - ./certs:/etc/nginx/certs
    depends_on:
     - syslog
    logging:
      driver: syslog
      options:
        syslog-address: tcp://127.0.0.1:10514
        tag: nginx
```

And the following result was shown.
```
$ docker-compose up
Creating nginx_syslog_1 ...
Creating nginx_syslog_1 ... done
Creating nginx_nginx_1 ...
Creating nginx_nginx_1 ... error

ERROR: for nginx_nginx_1  Cannot create container for service nginx: plugin hbm failed with error: AuthZPlugin.AuthZReq: an error occurred; contact your system administrator

ERROR: for nginx  Cannot create container for service nginx: plugin hbm failed with error: AuthZPlugin.AuthZReq: an error occurred; contact your system administrator
ERROR: Encountered errors while bringing up the project.
```
When `nginx` service run without syslog driver, no error happened.

I checked `HBM` log.
```
Recovered: runtime error: invalid memory address or nil pointer dereference
time="2018-04-24T17:41:50+09:00" level=warning msg="Recovered panic: runtime error: invalid memory address or nil pointer dereference"
time="2018-04-24T17:41:50+09:00" level=warning msg="goroutine 8 [running]:\nruntime/debug.Stack(0xc420321c40, 0xc420910b40, 0x2)\n\t/usr/local/go/src/runtime/debug/stack.go:24 +0x79\ngithub.com/kassisol/hbm/plugin.(*Api).Allow.func1(0xf894e0, 0xc420321c40, 0xf85540, 0xc420246940, 0xc4209e3950)\n\t/go/src/github.com/kassisol/hbm/plugin/api.go:56 +0x137\npanic(0xa453c0, 0xf6f0b0)\n\t/usr/local/go/src/runtime/panic.go:489 +0x2cf\ngithub.com/kassisol/hbm/plugin.(*Api).Allow(0xc420321c20, 0x0, 0x0, 0x0, 0x0, 0xc4200f5c30, 0x4, 0xc4203de930, 0x2b, 0xc4201b5400, ...)\n\t/go/src/github.com/kassisol/hbm/plugin/api.go:116 +0x862\ngithub.com/kassisol/hbm/plugin.(*plugin).AuthZReq(0xc420366cf0, 0x0, 0x0, 0x0, 0x0, 0xc4200f5c30, 0x4, 0xc4203de930, 0x2b, 0xc4201b5400, ...)\n\t/go/src/github.com/kassisol/hbm/plugin/plugin.go:52 +0x238\ngithub.com/kassisol/hbm/vendor/github.com/docker/go-plugins-helpers/authorization.(*Handler).initMux.func1(0x0, 0x0, 0x0, 0x0, 0xc4200f5c30, 0x4, 0xc4203de930, 0x2b, 0xc4201b5400, 0x4b9, ...)\n\t/go/src/github.com/kassisol/hbm/vendor/github.com/docker/go-plugins-helpers/authorization/api.go:118 +0xa0\ngithub.com/kassisol/hbm/vendor/github.com/docker/go-plugins-helpers/authorization.(*Handler).handle.func1(0xf849c0, 0xc4200d0380, 0xc420334500)\n\t/go/src/github.com/kassisol/hbm/vendor/github.com/docker/go-plugins-helpers/authorization/api.go:139 +0x14c\nnet/http.HandlerFunc.ServeHTTP(0xc42037ecc0, 0xf849c0, 0xc4200d0380, 0xc420334500)\n\t/usr/local/go/src/net/http/server.go:1942 +0x44\nnet/http.(*ServeMux).ServeHTTP(0xc420366de0, 0xf849c0, 0xc4200d0380, 0xc420334500)\n\t/usr/local/go/src/net/http/server.go:2238 +0x130\nnet/http.serverHandler.ServeHTTP(0xc420076f20, 0xf849c0, 0xc4200d0380, 0xc420334500)\n\t/usr/local/go/src/net/http/server.go:2568 +0x92\nnet/http.(*conn).serve(0xc42036d400, 0xf85280, 0xc42037b500)\n\t/usr/local/go/src/net/http/server.go:1825 +0x612\ncreated by net/http.(*Server).Serve\n\t/usr/local/go/src/net/http/server.go:2668 +0x2ce\n"
```

I already tried to run with various versions, `0.9.2`, `0.9.5`, `0.10.0`.
Looking at stacktrace, an error point is https://github.com/kassisol/hbm/blob/0.9.2/plugin/api.go#L116.
And then after executing https://github.com/kassisol/hbm/blob/0.9.2/plugin/api.go#L102, r was set to nil.

I don't know what I should do to debug at this point.

**Comment by juliengk** [comment]
Hi,

I was not able to reproduce the error. Do you have more logs? Could you paste the output of the commands `hbm resource ls` and `hbm policy ls` please?

Thank you

**Comment by odg0318** [comment]
@juliengk I found the exact error line.
https://github.com/kassisol/hbm/blob/0.9.2/storage/driver/sqlite/policy.go#L48
I printed the argument of the function, `ListPolicies`.
```golang
func (c *Config) ListPolicies(filter map[string]string) []types.Policy {
```
`filter` map consists of `user`: `all`, `resource-type`: `all`, `resource-value`: `all`, `resource-options`: `all`.
And I printed a log of the following.
```golang
	rows, err := sql.Rows()
         if err != nil {
                   println(err)
         }
	defer rows.Close()
```
The message shows `sql: database is closed`.

How can I dump sql query and what should I check?

**Comment by odg0318** [comment]
I found how to dump sql query. The query is below.

```
SELECT policies.name, groups.name, collections.name FROM "policies" JOIN groups ON groups.id = policies.group_id JOIN collections ON collections.id = policies.collection_id JOIN group_users ON group_users.group_id = groups.id JOIN users ON users.id = group_users.user_id JOIN collection_resources ON collection_resources.collection_id = collections.id JOIN resources ON resources.id = collection_resources.resource_id WHERE (users.name = 'all') AND (resources.type = 'all' AND resources.value = 'all') AND (resources.option = 'all')
```
It seems like that the query itself has no error. Database connection seems an error point.

**Comment by odg0318** [comment]
I found the reason; https://github.com/kassisol/hbm/blob/master/docker/allow/container.go#L139.
Seems like a mistake? Can I make new PR?

**Comment by juliengk** [comment]
yes please make a PR.

**Comment by odg0318** [comment]
Resolved #22
